load-gen/src/main/java/io/smartcat/data/loader/datasource/CSVDataSource.java

```diff
@@ -0,0 +1,95 @@
+package io.smartcat.data.loader.datasource;
+
+import java.util.Iterator;
+
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.smartcat.data.loader.api.DataSource;
+
+/**
+ * Data source that reads data from CSV file.
+ *
+ * @param <T> Type of the data provided by data source implementation.
+ */
+public class CSVDataSource<T> implements DataSource<T>, AutoCloseable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CSVDataSource.class);
+
+    private final CSVParser csvParser;
+    private final RowMapper<T> rowMapper;
+
+    private Iterator<CSVRecord> iterator;
+    private boolean closed = false;
+
+    /**
+     * Constructs CSV data source with specified <code>csvParser</code> and <code>rowMapper</code>.
+     *
+     * @param csvParser parser which provides CSV records to this data source.
+     * @param rowMapper mapper which maps CSV records to type {@code <T>}.
+     */
+    public CSVDataSource(CSVParser csvParser, RowMapper<T> rowMapper) {
+        if (csvParser == null) {
+            throw new IllegalArgumentException("CSV parser cannot be null.");
+        }
+        if (rowMapper == null) {
+            throw new IllegalArgumentException("Row mapper cannot be null.");
+        }
+        this.rowMapper = rowMapper;
+        this.csvParser = csvParser;
+        this.iterator = this.csvParser.iterator();
+    }
+
+    @Override
+    public boolean hasNext(long time) {
+        if (closed) {
+            throw new IllegalStateException("Data source is already closed.");
+        }
+        return iterator.hasNext();
+    }
+
+    @Override
+    public T getNext(long time) {
+        if (closed) {
+            throw new IllegalStateException("Data source is already closed.");
+        }
+        CSVRecord record = iterator.next();
```

Comment:
It cannot throw error. It might be hard to believe, but it will always split data to columns, it might just happen that number of columns is not the same as expected. Only error that it throws is NoSuchElementException when there is no more records in CSV. That exception is propagated. Should I handle it also?